velox/functions/prestosql/DecimalArithmetic.cpp

```diff
@@ -0,0 +1,249 @@
+      auto rawB = flatB->mutableRawValues();
+
+      R* rawResults = nullptr;
+      if (!(*result)) {
+        // Try to reuse one of the flat vectors to store result.
+        if (resultType->kind() == flatA->typeKind() &&
+            BaseVector::isReusableFlatVector(args[0])) {
+          rawResults = reinterpret_cast<R*>(rawA);
+          *result = std::move(args[0]);
+        } else if (
+            resultType->kind() == flatB->typeKind() &&
+            BaseVector::isReusableFlatVector(args[1])) {
+          rawResults = reinterpret_cast<R*>(rawB);
+          *result = std::move(args[1]);
+        }
+      }
+      if (!rawResults) {
+        rawResults = prepareResults(rows, resultType, context, result);
+      }
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawA[row], rawB[row], aRescale_, bRescale_);
+      });
+    } else {
+      exec::DecodedArgs decodedArgs(rows, args, context);
+      auto a = decodedArgs.at(0);
+      auto b = decodedArgs.at(1);
+      BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+      auto rawResults =
+          (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row],
+            a->valueAt<A>(row),
+            b->valueAt<B>(row),
+            aRescale_,
+            bRescale_);
+      });
+    }
+  }
+
+ private:
+  R* prepareResults(
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+    (*result)->clearNulls(rows);
+    return (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+  }
+
+  R* getRawResults(
+      const SelectivityVector& rows,
+      VectorPtr& flat,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    // Check if input can be reused for results.
+    R* rawResults;
+    if (!(*result) && flat->typeKind() == resultType->kind() &&
+        BaseVector::isReusableFlatVector(flat)) {
+      auto flatValues = flat->asUnchecked<FlatVector<R>>();
+      rawResults = flatValues->mutableRawValues();
+      *result = std::move(flat);
+    } else {
+      rawResults = prepareResults(rows, resultType, context, result);
+    }
+    return rawResults;
+  }
+  uint8_t aRescale_;
+  uint8_t bRescale_;
+};
+
+class Addition {
+ public:
+  template <typename R, typename A, typename B>
+  inline static void
+  apply(R& r, const A& a, const B& b, uint8_t aRescale, uint8_t bRescale) {
+    r.setUnscaledValue(
+        a.unscaledValue() * kPowersOfTen[aRescale] +
+        b.unscaledValue() * kPowersOfTen[bRescale]);
+  }
+
+  inline static uint8_t
+  computeRescaleFactor(uint8_t fromScale, uint8_t toScale, uint8_t rScale = 0) {
```

Comment:
Do we need rScale parameter?

Reply:
Yes. For computing rescale factor for dividend we require rScale. Refer:
https://prestodb.io/docs/current/functions/decimal.html#binary-arithmetic-decimal-operators